### PR TITLE
Use a special DeclBaseName for `init`

### PR DESCRIPTION
**Pull request by slavapestov**
Fixes <https://bugs.swift.org/browse/SR-1660>, <rdar://problem/35852727>, <https://bugs.swift.org/browse/SR-6557>.

### CODE REVIEW

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

File: lib/IRGen/GenProto.cpp [hunk 1]
```diff
@@ -1000,30 +999,6 @@ emitConditionalConformancesBuffer(IRGenFunction &IGF, CanType conformingType,
...
-    });
-
-    if (conformingTypeEnv) {
-      subMap = subMap.subst([&](SubstitutableType *dependentType) {
-            return conformingTypeEnv->mapTypeIntoContext(Type(dependentType));
-          },
-          LookUpConformanceInModule(IGF.getSwiftModule()));
-    }
-  }
-
```
Comment (jrose-apple) [inline lib/IRGen/GenProto.cpp]: All this IRGen stuff should probably go in another PR, even if it's correct.

Comment (slavapestov) [reply lib/IRGen/GenProto.cpp]: ... and apparently it doesn't compile. I'll tag you once its done.

File: lib/AST/ASTContext.cpp [hunk 2]
```diff
@@ -4289,6 +4289,9 @@ void DeclName::CompoundDeclName::Profile(llvm::FoldingSetNodeID &id,
 
 void DeclName::initialize(ASTContext &C, DeclBaseName baseName,
                           ArrayRef<Identifier> argumentNames) {
+  if (!baseName.isSpecial())
+    assert(baseName.getIdentifier() != C.getIdentifier("init"));
```
Comment (jrose-apple) [inline lib/AST/ASTContext.cpp]: Isn't this the very thing that will break with `` `init` ``?

Comment (slavapestov) [reply lib/AST/ASTContext.cpp]: The assertion is only temporary while I work out some other issues.

**Comment by slavapestov** [comment]
@swift-ci Please test source compatibility

**Comment by slavapestov** [comment]
@swift-ci Please test source compatibility

**Comment by slavapestov** [comment]
@swift-ci Please test

**Comment by slavapestov** [comment]
@swift-ci Please test source compatibility

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please test source compatibility

**Comment by slavapestov** [comment]
@swift-ci Please test

**Comment by slavapestov** [comment]
@swift-ci Please test source compatibility

**Comment by slavapestov** [comment]
@swift-ci Please test
